### PR TITLE
Try to detect sha2 from libmd

### PR DESCRIPTION
**Pull request by paravoid**
libmd provides SHA2 functions as well, so try to check if the system
copy provides them by passing -lmd, in the exact same manner as is
currently implemented for MD5.

### CODE REVIEW

**Comment by kristapsdz** [comment]
Yikes!  This is an ugly part of oconfigure.  I'm currently (1) moving all `HAVE_SHA2_H` into `HAVE_SHA2`, using your "-lmd" fix, and also properly documenting this in the README.

**Comment by kristapsdz** [comment]
Ok, the "-lmd" parts don't work and now I remember why: FreeBSD has a slightly different API than the one provided by OpenBSD.

**Comment by paravoid** [comment]
Is this the e.g. `SHA256_Init` vs. `SHA256Init` difference? Wouldn't the test fail in that case?

To complicate matters further... my intention was to use this with https://www.hadrons.org/software/libmd/ (present in e.g. Debian) which seems to be based on code from NetBSD/OpenBSD and thus follow their API rather than FreeBSD's. At least oconfigure's test works with this variant, although I haven't tested any actual software using it.

**Comment by paravoid** [comment]
@kristapsdz Thoughts on the above? Sorry for bugging you, but given the PR is closed I fear that this will get forgotten :)

**Comment by kristapsdz** [comment]
I think this is just complicating oconfigure: the point of the shim is to simply see whether the base system provides the given OpenBSD-style functions; and if not, and if possible, to provide a substitute.  It will currently fail with hadrons' libmd, as with libbsd, because these require namespaced headers---even if the functions are named the same.  So the test won't pick up the namespaced functions.  (This is probably obvious to you, but I'm making sure we're on the same page.)

The test *would* pick up "overlay" functions, so long as the overlay is specified in `CFLAGS`, `CPPFLAGS`, and the appropriate libraries given in `LDADD` and `LDFLAGS`.

However, specifically regarding SHA2, what's the point, since oconfigure provides its own?  Doesn't that mean you don't need libmd installed at all?

**Comment by paravoid** [comment]
> I think this is just complicating oconfigure: the point of the shim is to simply see whether the base system provides the given OpenBSD-style functions; and if not, and if possible, to provide a substitute. It will currently fail with hadrons' libmd, as with libbsd, because these require namespaced headers---even if the functions are named the same. So the test won't pick up the namespaced functions. (This is probably obvious to you, but I'm making sure we're on the same page.)

This is actually not the case with hadron's libmd. Example from a [Debian system](https://packages.debian.org/buster/amd64/libmd-dev/filelist):
```shell
$ dpkg-query -W libmd-dev
libmd-dev:amd64	1.0.3-1
$ dpkg -L libmd-dev | grep sha2.h
/usr/include/sha2.h
$ grep SHA256Init /usr/include/sha2.h
void SHA256Init(SHA2_CTX *)
```
IOW, in the oconfigure source tree, `gcc test-sha2.c -lmd` works. With the commit in the PR here, `oconfigure` succeds in detecting the system copy.

So libmd in this case provides the given OpenBSD-style functions, as you said, and can be used instead of the embedded copy for both MD5 and SHA2. oconfigure currently does so for MD5, but not SHA2.

In other words, hadron's libmd is different than FreeBSD's libmd: it's based on OpenBSD' code and exposes primarily the OpenBSD API (but is [also compatible](https://git.hadrons.org/cgit/libmd.git/plain/include/sha256.h) with the FreeBSD one.)

(Since you mentioned libbsd: libbsd does not provide SHA2 functions. It only provides the MD5 ones, and these are marked deprecated and to be removed in a later version, in favor of libmd. The two libraries are from the same author (hadrons/guillemj), so I suppose at some point he decided to spin-off crypto into a separate library.)

Thanks for trying to make sure we were on the same page; evidently that was helpful! :)

**Comment by kristapsdz** [comment]
Oh, I didn't know that it was directly in /usr/include.  So if I have the "-lmd" bits set for the SHA2 test, it does pick up the hadron md library correctly?

**Comment by paravoid** [comment]
That's correct :)

**Comment by kristapsdz** [comment]
I've integrated your patch and documented it.  Please test it, and if it works, close this commit.  I'll then tag out a release and we can start to look in detail at libbsd.  Thanks again for reopening this!

**Comment by paravoid** [comment]
The detection works, but the regression test suite wasn't adjusted and so it failed (that was my bad!). Diff:
```diff
--- a/Makefile
+++ b/Makefile
@@ -7,6 +7,7 @@ REGRESS_MD5		= regress/md5
 REGRESS_CRYPT		= regress/crypt
 REGRESS_B64		= regress/b64_ntop
 REGRESS_LIB_SOCKET	= regress/SOCK_NONBLOCK
+REGRESS_SHA2		= regress/sha2
 REGRESS_NODEP		= regress/capsicum \
 			  regress/endian \
 			  regress/err \
@@ -25,7 +26,6 @@ REGRESS_NODEP		= regress/capsicum \
 			  regress/recallocarray \
 			  regress/setresgid \
 			  regress/setresuid \
-			  regress/sha2 \
 			  regress/strndup \
 			  regress/strnlen \
 			  regress/strlcpy \
@@ -38,6 +38,7 @@ REGRESS			= $(REGRESS_B64) \
 			  $(REGRESS_CRYPT) \
 			  $(REGRESS_LIB_SOCKET) \
 			  $(REGRESS_MD5) \
+			  $(REGRESS_SHA2) \
 			  $(REGRESS_NODEP)
 
 all: $(REGRESS)
@@ -60,6 +61,11 @@ ${r}: ${r}.c compats.o config.h
 	$(CC) $(CFLAGS) $(CPPFLAGS) -o $@ ${r}.c compats.o $(LDADD_MD5)
 .endfor
 
+.for r in $(REGRESS_SHA2)
+${r}: ${r}.c compats.o config.h
+	$(CC) $(CFLAGS) $(CPPFLAGS) -o $@ ${r}.c compats.o $(LDADD_SHA2)
+.endfor
+
 .for r in $(REGRESS_B64)
 ${r}: ${r}.c compats.o config.h
 	$(CC) $(CFLAGS) $(CPPFLAGS) -o $@ ${r}.c compats.o $(LDADD_B64_NTOP)
```

(btw, `make` with GNU Make fails -- I had to run this with `apt install bmake; bmake regress`)

**Comment by kristapsdz** [comment]
Done!  I don't have any place documenting "make regress", so I don't really mention it.  If you'd like to add a "contributing" section to the README.md with this, then by all means do so!  I've tagged out 0.3.1 and can now look more fully at libbsd.  Thanks!
